### PR TITLE
feat: add stop orders protobuf and domain types

### LINKED ISSUE #8466: Stop orders command API
Implement the new command for stop orders.

To be covered:
- [] New proto command
- [] New proto model
- [] Validation
- [] wired into the processor

### PR DESCRIPTION
**Pull request by jeremyletang**
close #8466

### CODE REVIEW

File: protos/sources/vega/commands/v1/commands.proto [hunk 1]
```diff
@@ -24,6 +24,54 @@ message BatchMarketInstructions {
...
   repeated OrderSubmission submissions = 3;
+  // List of stop orders submissions to be processed sequentially.
+  repeated StopOrdersSubmission stop_orders_submissions = 4;
+}
+
+// Stop order submission submits stops orders
+// It is possible to make a single stop order submission by
+// specifying a single direction,
+// or an OCO (One Cancels the Other) stop order submission
+// by specifying a configuration for both direction
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:35]:
```suggestion
// Stop order submission submits stops orders.
// It is possible to make a single stop order submission by
// specifying a single direction,
// or an OCO (One Cancels the Other) stop order submission
// by specifying a configuration for both directions
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 2]
```diff
@@ -24,6 +24,54 @@ message BatchMarketInstructions {
...
+message StopOrdersSubmission {
+  // A stop order which will be trigger
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:37]:
```suggestion
  // Stop order that will be triggered
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 3]
```diff
@@ -24,6 +24,54 @@ message BatchMarketInstructions {
...
+// The price and expiry configuration for a stop order
+message StopOrderSetup {
+  // The order to be submitted once the trigger is breached.
+  OrderSubmission order_submission = 1;
+  // An optional expiry timestamp.
+  optional int64 expires_at = 2;
+  // The strategy to adopt if the expiry time is reached.
+  optional vega.StopOrder.ExpiryStrategy expiry_strategy = 3;
+
+  // The trigger which will need to be breached for the order
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:54]:
```suggestion
  // Trigger that will need to be breached for the order
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 4]
```diff
@@ -24,6 +24,54 @@ message BatchMarketInstructions {
...
+  // to be submitted to the book.
+  oneof trigger {
+    // A fixed price at which the order will be submitted.
+    string price = 100;
+    // A trailing percentage at wich the order will be submitted.
+    string trailing_percent_offset = 101;
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:60]:
```suggestion
    // Fixed price at which the order will be submitted.
    string price = 100;
    // Trailing percentage at which the order will be submitted.
    string trailing_percent_offset = 101;
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 5]
```diff
@@ -24,6 +24,54 @@ message BatchMarketInstructions {
...
+  }
+}
+
+// Cancel a stop order.
+// the following combination are available:
+// an empty object will cancel all stop orders for the party
+// a market id alone, will cancel all stop orders in a market
+// a market id and order id will cancel a specific stop order in a market
+// If the stop order is part of an OCO, both stop order will be cancelled.
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:69]:
```suggestion
// Cancel a stop order.
// The following combinations are available:
// Empty object will cancel all stop orders for the party
// Market ID alone will cancel all stop orders in a market
// Market ID and order ID will cancel a specific stop order in a market
// If the stop order is part of an OCO, both stop orders will be cancelled
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 6]
```diff
@@ -24,6 +24,54 @@ message BatchMarketInstructions {
...
+message StopOrdersCancellation {
+  // An optional market id.
+  optional string market_id = 1;
+  // An optional stop_order_id.
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:73]:
```suggestion
  // Optional market ID.
  optional string market_id = 1;
  // Optional order ID.
```

Is this order ID unique to stopped orders or is it the ID of the order that should be stopped? I've changed it assuming it's the order that should be stopped.

File: protos/sources/vega/vega.proto [hunk 7]
```diff
@@ -6,6 +6,73 @@ import "vega/markets.proto";
 
 option go_package = "code.vegaprotocol.io/vega/protos/vega";
 
+message StopOrder {
+  enum ExpiryStrategy {
+    // Never valid
+    EXPIRY_STRATEGY_UNSPECIFIED = 0;
+    // The stop order should be cancelled if the expiry time is reached.
+    EXPIRY_STRATEGY_CANCELS = 1;
+    // The order should be submitted if the expiry time is reached.
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:15]:
```suggestion
    // Order should be submitted if the expiry time is reached.
```

File: protos/sources/vega/vega.proto [hunk 8]
```diff
@@ -6,6 +6,73 @@ import "vega/markets.proto";
...
+  }
+
+  enum Status {
+    // Never valid
+    STATUS_UNSPECIFIED = 0;
+    // Pending to be executed once the trigger is breached
+    STATUS_PENDING = 1;
+    // Cancelled by the user
+    STATUS_CANCELLED = 2;
+    // Stopped by the network, e.g: OCO other side has been triggered
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:35]:
```suggestion
    // Stopped by the network, e.g: OCO on the other side has been triggered
```

File: protos/sources/vega/vega.proto [hunk 9]
```diff
@@ -6,6 +6,73 @@ import "vega/markets.proto";
...
+    STATUS_STOPPED = 3;
+    // Stop order has been triggered and generated an order
+    STATUS_TRIGGERRED = 4;
+    // Stop order has been expired
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:39]:
```suggestion
    // Stop order has expired
```

File: protos/sources/vega/vega.proto [hunk 10]
```diff
@@ -6,6 +6,73 @@ import "vega/markets.proto";
...
+    STATUS_EXPIRED = 5;
+  }
+
+  // The ID of order of this stop order
+  // also the ID of the associated order if it is ever trigger
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:44]:
```suggestion
  // ID of this stop order
  // also the ID of the associated order if it is ever triggered
```

File: protos/sources/vega/vega.proto [hunk 11]
```diff
@@ -6,6 +6,73 @@ import "vega/markets.proto";
...
+  string id = 1;
+  // The id of the other part of the OCO if 2 stop orders where submitted at once
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:46]:
```suggestion
  // The ID of the 'other' part of the OCO if 2 stop orders were submitted at once
```

File: protos/sources/vega/vega.proto [hunk 12]
```diff
@@ -6,6 +6,73 @@ import "vega/markets.proto";
...
+  // Status of the stop order
+  Status status = 6;
+  // Creation time of the stop order
+  int64 created_at = 7;
+  // Last update of this stop order
+  optional int64 updated_at = 8;
+  // The ID of the order created once the trigger is hit
+  string order_id = 9;
+  // The order to be submitted once the trigger is hit.
+  // OrderSubmission order_submission = 10;
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:63]:
```suggestion
  // Optional expiry timestamp.
  optional int64 expires_at = 3;
  // Strategy to adopt if the expiry time is reached.
  optional ExpiryStrategy expiry_strategy = 4;
  // Trigger direction for this stop order.
  optional TriggerDirection trigger_direction = 5;
  // Status of the stop order.
  Status status = 6;
  // Creation time of the stop order.
  int64 created_at = 7;
  // Last update of this stop order.
  optional int64 updated_at = 8;
  // ID of the order created once the trigger is hit.
  string order_id = 9;
  // Order to be submitted once the trigger is hit.
  // OrderSubmission order_submission = 10;
```

File: protos/sources/vega/vega.proto [hunk 13]
```diff
@@ -6,6 +6,73 @@ import "vega/markets.proto";
...
+
+
+  // The trigger which will need to be breached for the order
+  // to be submitted to the book.
+  oneof trigger {
+    // A fixed price at which the order will be submitted.
+    string price = 100;
+    // A trailing percentage at wich the order will be submitted.
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:71]:
```suggestion
  // Trigger that will need to be breached for the order
  // to be submitted to the book.
  oneof trigger {
    // Fixed price at which the order will be submitted.
    string price = 100;
    // Trailing percentage at which the order will be submitted.
```

File: protos/sources/vega/commands/v1/transaction.proto [hunk 14]
```diff
@@ -68,6 +68,10 @@ message InputData {
     AnnounceNode announce_node = 1014;
     // Command to submit a batch of order instructions to a market.
     BatchMarketInstructions batch_market_instructions = 1015;
+    // Command to submit a pair of stop orders
+    StopOrdersSubmission stop_orders_submission = 1016;
+    // Command to cancel stop orders
+    StopOrdersCancellation stop_orders_cancellation = 1017;
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/transaction.proto:74]:
```suggestion
    // Command to submit a pair of stop orders.
    StopOrdersSubmission stop_orders_submission = 1016;
    // Command to cancel stop orders.
    StopOrdersCancellation stop_orders_cancellation = 1017;
```

File: core/types/stop_orders.go [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+	StopOrderTriggerDirectionRisesAbove = vega.StopOrder_TRIGGER_DIRECTION_RISES_ABOVE
+)
+
+type StopOrderStatus = vega.StopOrder_Status
+
+const (
+	// Never valid.
+	StopOrderStatusUnspecified StopOrderStatus = vega.StopOrder_STATUS_UNSPECIFIED
+	// Pending to be executed once the trigger is breached.
+	StopOrderStatusPending = vega.StopOrder_STATUS_CANCELLED
```
Comment (wwestgarth) [inline core/types/stop_orders.go:42]:
```suggestion
	StopOrderStatusPending = vega.StopOrder_STATUS_PENDING
```

Comment (jeremyletang) [reply core/types/stop_orders.go:42]: good one.

Comment (jeremyletang) [reply core/types/stop_orders.go]: addressed

**Review by wwestgarth — APPROVED** [review]

**Pull request merged**